src/contract.ts

```diff
@@ -0,0 +1,381 @@
+
+    const contractStorageAccount = await program.createStorageAccount(space);
+    const abi = new ethers.utils.Interface(contractAbiData);
+    const input = abi.encodeDeploy(constructorArgs);
+
+    let hash = ethers.utils.keccak256(Buffer.from(contractName));
+
+    const data = Buffer.concat([
+      contractStorageAccount.publicKey.toBuffer(),
+      caller.toBuffer(),
+      Buffer.from(numToPaddedHex(value), 'hex'),
+      Buffer.from(hash.substr(2, 8), 'hex'),
+      encodeSeeds(seeds),
+      Buffer.from(input.replace('0x', ''), 'hex'),
+    ]);
+
+    const keys = [
+      // ...seeds.map((seed) => ({
+      //   pubkey: seed.address,
+      //   isSigner: false,
+      //   isWritable: true,
+      // })),
+      {
+        pubkey: contractStorageAccount.publicKey,
+        isSigner: false,
+        isWritable: true,
+      },
+      // {
+      //   pubkey: SYSVAR_CLOCK_PUBKEY,
+      //   isSigner: false,
+      //   isWritable: false,
+      // },
+      // {
+      //   pubkey: PublicKey.default,
+      //   isSigner: false,
+      //   isWritable: false,
+      // },
+      ...accounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: false,
+      })),
+      ...writableAccounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: true,
+      })),
+    ];
+
+    signers.unshift(program.payerAccount);
+
+    const instruction = new TransactionInstruction({
+      keys,
+      programId: program.programAccount.publicKey,
+      data,
+    });
+
+    const { logs, computeUnitsUsed } = await program.makeTx(
+      simulate,
+      instruction,
+      signers
+    );
+
+    const contract = new Contract(
+      program,
+      contractStorageAccount,
+      contractAbiData
+    );
+
+    const events = contract.parseLogsEvents(logs);
+
+    return {
+      contract,
+      logs,
+      computeUnitsUsed,
+      events,
+    };
+  }
+
+  /**
+   * Load a deployed contract
+   *
+   * @param program
+   * @param abiData
+   * @param contractStorageAccount
+   * @returns
+   */
+  static async get(
+    program: Program,
+    abiData: string,
+    contractStorageAccount: Keypair
+  ): Promise<Contract> {
+    return new Contract(program, contractStorageAccount, abiData);
+  }
+
+  public abi: ethers.utils.Interface;
+  readonly functions: { [name: string]: ContractFunction };
+
+  /**
+   *
+   * @param program
+   * @param contractStorageAccount
+   * @param abiData
+   */
+  constructor(
+    public program: Program,
+    public contractStorageAccount: Keypair,
+    public abiData: string
+  ) {
+    this.abi = new ethers.utils.Interface(abiData);
+    this.functions = {};
+    Object.entries(this.abi.functions).forEach(([, frag]) => {
+      this.functions[frag.name] = this.buildCall(frag);
+    });
+  }
+
+  /**
+   *
+   * @param fragment
+   * @returns
+   */
+  buildCall(fragment: ethers.utils.Fragment): ContractFunction {
+    return (...args: Array<any>) => {
```

Comment:
What could be done to infer the types here? If the API for every function is effectively
```ts
(...args: any[]) => Promise<any>
```
this pushes a lot of complexity into the application.

Reply:
@jordansexton what do you think?

Reply:
Yeah. The codegen of TypeScript when using hardhat-ethers is **really** nice

Reply:
gotcha. looks like typechain still uses ethers underneath https://github.com/dethcrypto/TypeChain/blob/7999f466164b7045a8c57e1fd92c667e24c6a71a/packages/target-ethers-v4-test/test/common.ts#L20

Reply:
I think we can postpone this for now..